### ISSUE TITLE
Introduce BisectionSolver for mixed precision quantization.

### ISSUE THREAD

**Issue by stamalakhov** [Samsung/ONE#10154]
### What

Introduce light-weight fast methods for finding quasi-optimal MP quantization. So let's start with `bisectioning` method.
### Why
We may need to build mixed precision quantized model in acceptable time aiming to achieve `admissible accuracy` with `reduced execution time`.

**Comment by jinevening** [comment]
Lightweight MPQ will definitely improve the usability of our MPQ feature :) Could you explain the `quasi-optimal` aspect of `bisection` method?

**Comment by stamalakhov** [comment]
> Lightweight MPQ will definitely improve the usability of our MPQ feature :) Could you explain the `quasi-optimal` aspect of `bisection` method?

`Bisection` assumes:
1) The `accuracy` can be parameterized with just a single parameter, e.g., its `depth` (`distance` from inputs nodes)
2) The `accuracy` is monotone with the neural network `depth`.
In this case `bisection` can provide solution in ~`ln(N) * T(inference)` (`N`- is maximal NN depth, `T(inference)` - time of full inference).

In general, these assumptions do not hold. Since NNs can be very heterogeneous they cannot be treated as one-dimensional structures. Moreover, initial training can be done using strong augmentation, which helps a lot to deal with quantization errors, so `accuracy` may be not monotone along NN's depth as well.
Summing up:
1) (speed) `bisection` - is fast enough
2) (`quasi-optimal` or `suboptimal`) but it may turn more Q8 layers into Q16 than needed to meet the `accuracy` requirement

**Comment by stamalakhov** [comment]
@jinevening
Do we need to implement calling of circle-mpqsolver via onecc interface or direct calling is ok?
I mean for UI implementation in ONE-vscode it would be convenient to use onecc.

**Comment by jinevening** [comment]
> Do we need to implement calling of circle-mpqsolver via onecc interface or direct calling is ok?

For now, direct calling is ok. `circle-mpqsolver` may be integrated with `onecc` when it is ready (working in more general scenarios). 

Do you have a plan to integrate circle-mpqsolver into ONE-vscode?  Could you describe your plan? Or it would be good to have a meeting about that item :)

**Comment by stamalakhov** [comment]
> For now, direct calling is ok. `circle-mpqsolver` may be integrated with `onecc` when it is ready (working in more general scenarios).

Ok.



> Do you have a plan to integrate circle-mpqsolver into ONE-vscode? Could you describe your plan? Or it would be good to have a meeting about that item :)

May be the following plan will be possible?
1) When 'auto' mode will be merged, 
2) it will be convenient to make it callable from the outside, (e.g. include it some toolchain), 
3) include GUI for it in ONE-vscode (e.g. on "Quantization" tab or add another tab)

A meeting about that option would be great )

**Comment by stamalakhov** [comment]
Seems like we can close this one, as preliminary version is ready, so further items will be processed by their own (more specific) issues.

**Comment by stamalakhov** [comment]
As a summary:
running `bisection` for `mobile_net_v1` (see attachment) on 100 images lasted for ~8 minutes and
for qerror_ratio=0.5 (target_error = 0.000202348) found optimal_depth=5 and produced `new_model.mpq.circle` (see attachment), while for qerror_ratio=0.1 (target_error=0.000126946) found optimal_depth=25 and produced `new_model_q16.mpq.circle` (see attachment).
[bisection_inputs_and_results.zip](https://github.com/Samsung/ONE/files/11798471/bisection_inputs_and_results.zip)

To produce `visq` data for `bisection` we used `visq` component on the same `h5` file. It took 3 minutes.
